libtsuba/include/tsuba/ReadGroup.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KATANA_LIBTSUBA_TSUBA_READGROUP_H_
+#define KATANA_LIBTSUBA_TSUBA_READGROUP_H_
+
+#include <future>
+#include <list>
+#include <memory>
+
+#include "katana/Result.h"
+#include "tsuba/AsyncOpGroup.h"
+
+namespace arrow {
+class Table;
+};
+
+namespace tsuba {
+
+/// Track multiple, outstanding async writes and provide a mechanism to ensure
+/// that they have all completed
+class ReadGroup {
+public:
+  static katana::Result<std::unique_ptr<ReadGroup>> Make();
+
+  /// Wait until all operations this descriptor knows about have completed
+  katana::Result<void> Finish();
+
+  /// Add future to the list of futures this ReadGroup will wait for, note
+  /// the file name for debugging. `on_complete` is guaranteed to be invoked
+  /// after `on_complete`'s passed by previous calls to AddOp
+  void AddOp(
+      std::future<katana::Result<void>> future, std::string file,
+      const std::function<katana::Result<void>()>& on_complete);
+
+  /// same as AddOp, but the future may return a data type which can then be
+  /// consumed by on_complete
+  void AddReturnsOp(
+      std::future<katana::Result<std::shared_ptr<arrow::Table>>> future,
+      const std::string& file,
+      const std::function<katana::Result<void>(std::shared_ptr<arrow::Table>)>&
+          on_complete) {
+    // n.b., make shared instead of unique because move capture below prevents
+    // passing generic_complete_fn as a std::function
+    auto ret_val = std::make_shared<std::shared_ptr<arrow::Table>>();
+    auto new_future = std::async(
+        std::launch::deferred,
+        [future = std::move(future),
+         &ret_val_storage = *ret_val]() mutable -> katana::Result<void> {
+          auto res = future.get();
```

Comment:
Actually, I have a question.  I'd generally think to prefer `async `to `deferred `for IO.  Oh, but this operation only gets called on a return, so we defer it?

Reply:
Yeah there's no real work in this function so it seemed fine to defer to call time when I was writing it;  probably it doesn't matter.

Reply:
Doing some research I think `std::launch::async` will always create a new thread for the current version of GCC (no pooling at all seems to be the consensus), so a good practice is probably to try and used `std::launch::deferred` as much as possible to avoid thread creation overheads; especially for no-work tasks like this one.

Reply:
Good to know.  A real shame they don't support thread pools.